### ISSUE TITLE
[2012 NIPS Spotlight] ImageNet Classification with Deep Convolutional Neural Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#1]
## 1. Introduction

- Appearance of Large Dataset
  * ImageNet consists of over 15 million labeled high-resolution images in over 22,000 categories
  * To learn about thousands of objects from millions of images, model must have large learning capacity and also have lots of prior knowledge to compensate for the data not on dataset
  
- Convolutional Neural Network
  * Capacity can be controlled by varying their depth and breadth
  * Make strong and mostly correct assumptions by stationarity of statistics and locality of pixel dependencies
  * Easier to train than standard feedforward neural networks cause of much fewer connections and parameters while theoretically-best performance is likely to be only slightly worse
  * Prohibitively expensive to apply in large scale to high-resolution images

- Current GPU
  * Powerful enough to facilitate the training of interestingly-large CNNs

- Contributions
  * Trained one of the largest convolutional neural networks using ILSVRC-2010, ILSVRC-2012 and achieved best results
  * Wrote highly-optimized GPU implementation of 2D convolution and all the operations inherent in training convolutional neural networks

- Training Time
  * 5~6 days to train on 2 GTX 580 3GB GPUs

## 2. The Dataset
- ImageNet
  * Consists of variable-resolution images
  * System requires constant input dimensionality
  * Down-sampled the images of a fixed resolution of 256 * 256
    * Rescale the image such that the shorter side was of length 256
    * Cropped out the central 256 * 256 patch
  * Image pre-processing
    * Subtract mean activity over the training set from each pixel

### 3.1. ReLU Nonlinearity
- Train several times faster than their equivalents with tanh units

### 3.2. Training on Multiple GPUs
- Spread the net across 2 GPUs cause of model size is too big to fit on one GPU

### 3.3. Local Response Normalization
- ReLUs do not require input normalization to prevent from saturating
- However local response normalization aids generalization
- Applied normalization after applying the ReLU nonlinearity in certain layers
- Can be seem as 'brightness normalization'

### 3.4. Overlapping Pooling
- Pooling layers in CNNs summarize the outputs of neighboring groups of neurons in the same kernel map
- Overlapping pooling is slightly more difficult to overfit

### 3.5. Overall Architecture
|Layer|Conv/Pool/FC Info|Input Size|Output Size|
|:-:|:-:|:-:|:-:|
|Conv1 + ReLU + LRN|(3, 96, 11, 4, 2, 1)|(B, 3, 224, 224)|(B, 96, 55, 55)|
|Pool1|(3, 2, 0)|(B, 96, 55, 55)|(B, 96, 27, 27)|
|Conv2 + ReLU + LRN|(96, 256, 5, 1, 2, 2)|(B, 96, 27, 27)|(B, 256, 27, 27)|
|Pool2|(3, 2, 0)|(B, 256, 27, 27)|(B, 256, 13, 13)|
|Conv3 + ReLU|(256, 384, 3, 1, 1, 1)|(B, 256, 13, 13)|(B, 384, 13, 13)|
|Conv4 + ReLU|(384, 384, 3, 1, 1, 2)|(B, 384, 13, 13)|(B, 384, 13, 13)|
|Conv5 + ReLU|(384, 256, 3, 1, 1, 2)|(B, 384, 13, 13)|(B, 256, 13, 13)|
|Pool3|(3, 2, 0)|(B, 256, 13, 13)|(B, 256, 6, 6)|
|FC1 + ReLU + Dropout|(256 * 6 * 6, 4096)|(B, 256 * 6 * 6)|(B, 4096)|
|FC2 + ReLU + Dropout|(4096, 4096)|(B, 4096)|(B, 4096)|
|FC3 + Softmax|(4096, 1000)|(B, 4096)|(B, 1000)|
- Conv layer : in_channels, out_channels, kernel_size, stride, padding, groups
- Pool layer : kernel_size, stride, padding
- FC layer : in_features, out_features
- B for batch size on Input Size and Output Size

### 4.1. Data Augmentation
- Generating Image Translations and Horizontal Reflections
  * Extract random 224 * 224 patches from 256 images on train (image can be horizontal reflected)
  * Extract 10 224 * 224 patches from 256 images on test (4 corner patches and center patch + horizontal reflections)
  * Average predictions of 10 patches from network's softmax layer on test

- Alter Intensity of the RGB Channels
  * Only applied on training images
  * PCA on the set of RGB pixel values through the training set
  * Object identity is invariant to changes in the intensity and color of the illumination

### 4.2. Dropout
- Reduces complex co-adaptations of neurons, since a neuron cannot rely on the presence of particular other neurons
- Forced to learn more robust features that are useful in conjunction with many different random subsets of the other neurons

## 5. Details of learning
- Stochastic Gradient Descent
  * Batch size 128
  * Momentum 0.9
  * Weight decay 0.0005

- Layer Initialization
  * Weight : zero-mean Gaussian distribution with standard deviation 0.01
  * Bias 
    * Constant 0 : conv 1, 3, fc 3
    * Constant 1 : conv 2, 4, 5 and fc 1, 2

- Learning Rate and Scheduler
  * Initial learning rate : 0.01
  * Divide by 10 when validation error rate stopped improving with current learning rate

## 6. Results
![image](https://user-images.githubusercontent.com/49643709/109617584-e947fb00-7b79-11eb-88e2-e1923a5fd437.png)
![image](https://user-images.githubusercontent.com/49643709/109617607-f06f0900-7b79-11eb-914a-18a3464b8faa.png)
- Models with asterisk * on ILSVRC-2012 : extra sixth convolutional layer over the last pooling layer
- 5 CNNs, 7 CNNs on ILSVRC-2012 : average the predictions

## 7. Discussion
- Deep convolutional neural network is capable of achieving record-breaking results on a highly challenging dataset using purely supervised learning
- Depth is important for achieving the results

**Comment by Jasonlee1995** [comment]
![image](https://user-images.githubusercontent.com/49643709/109740485-abe17d00-7c0e-11eb-8157-e28d0b352d84.png)

- I don't know exactly how authors get theoretically-best performance of standard feedforward neural network and CNN, but VC dimension may be the key for measuring the upper-bound of them
- [[VC Dimension wikipedia]](https://en.wikipedia.org/wiki/Vapnik%E2%80%93Chervonenkis_dimension)

**Comment by Jasonlee1995** [comment]
# Paper Summary

## 1. Introduction

- Appearance of Large Dataset
  * ImageNet consists of over 15 million labeled high-resolution images in over 22,000 categories
  * To learn about thousands of objects from millions of images, model must have large learning capacity and also have lots of prior knowledge to compensate for the data not on dataset
  
- Convolutional Neural Network
  * Capacity can be controlled by varying their depth and breadth
  * Make strong and mostly correct assumptions by stationarity of statistics and locality of pixel dependencies
  * Easier to train than standard feedforward neural networks cause of much fewer connections and parameters while theoretically-best performance is likely to be only slightly worse
  * Prohibitively expensive to apply in large scale to high-resolution images

- Current GPU
  * Powerful enough to facilitate the training of interestingly-large CNNs

- Contributions
  * Trained one of the largest convolutional neural networks using ILSVRC-2010, ILSVRC-2012 and achieved best results
  * Wrote highly-optimized GPU implementation of 2D convolution and all the operations inherent in training convolutional neural networks

- Training Time
  * 5~6 days to train on 2 GTX 580 3GB GPUs


## 2. The Dataset
- ImageNet
  * Consists of variable-resolution images
  * System requires constant input dimensionality
  * Down-sampled the images of a fixed resolution of 256 * 256
    * Rescale the image such that the shorter side was of length 256
    * Cropped out the central 256 * 256 patch
  * Image pre-processing
    * Subtract mean activity over the training set from each pixel


## 3. The Architecture
### 3.1. ReLU Nonlinearity
- Train several times faster than their equivalents with tanh units

### 3.2. Training on Multiple GPUs
- Spread the net across 2 GPUs cause of model size is too big to fit on one GPU

### 3.3. Local Response Normalization
- ReLUs do not require input normalization to prevent from saturating
- However local response normalization aids generalization
- Applied normalization after applying the ReLU nonlinearity in certain layers
- Can be seem as 'brightness normalization'

### 3.4. Overlapping Pooling
- Pooling layers in CNNs summarize the outputs of neighboring groups of neurons in the same kernel map
- Overlapping pooling is slightly more difficult to overfit

### 3.5. Overall Architecture
|Layer|Conv/Pool/FC Info|Input Size|Output Size|
|:-:|:-:|:-:|:-:|
|Conv1 + ReLU + LRN|(3, 96, 11, 4, 2, 1)|(B, 3, 224, 224)|(B, 96, 55, 55)|
|Pool1|(3, 2, 0)|(B, 96, 55, 55)|(B, 96, 27, 27)|
|Conv2 + ReLU + LRN|(96, 256, 5, 1, 2, 2)|(B, 96, 27, 27)|(B, 256, 27, 27)|
|Pool2|(3, 2, 0)|(B, 256, 27, 27)|(B, 256, 13, 13)|
|Conv3 + ReLU|(256, 384, 3, 1, 1, 1)|(B, 256, 13, 13)|(B, 384, 13, 13)|
|Conv4 + ReLU|(384, 384, 3, 1, 1, 2)|(B, 384, 13, 13)|(B, 384, 13, 13)|
|Conv5 + ReLU|(384, 256, 3, 1, 1, 2)|(B, 384, 13, 13)|(B, 256, 13, 13)|
|Pool3|(3, 2, 0)|(B, 256, 13, 13)|(B, 256, 6, 6)|
|FC1 + ReLU + Dropout|(256 * 6 * 6, 4096)|(B, 256 * 6 * 6)|(B, 4096)|
|FC2 + ReLU + Dropout|(4096, 4096)|(B, 4096)|(B, 4096)|
|FC3 + Softmax|(4096, 1000)|(B, 4096)|(B, 1000)|
- Conv layer : in_channels, out_channels, kernel_size, stride, padding, groups
- Pool layer : kernel_size, stride, padding
- FC layer : in_features, out_features
- B for batch size on Input Size and Output Size


## 4. Reducing Overfitting
### 4.1. Data Augmentation
- Generating Image Translations and Horizontal Reflections
  * Extract random 224 * 224 patches from 256 images on train (image can be horizontal reflected)
  * Extract 10 224 * 224 patches from 256 images on test (4 corner patches and center patch + horizontal reflections)
  * Average predictions of 10 patches from network's softmax layer on test

- Alter Intensity of the RGB Channels
  * Only applied on training images
  * PCA on the set of RGB pixel values through the training set
  * Object identity is invariant to changes in the intensity and color of the illumination

### 4.2. Dropout
- Reduces complex co-adaptations of neurons, since a neuron cannot rely on the presence of particular other neurons
- Forced to learn more robust features that are useful in conjunction with many different random subsets of the other neurons


## 5. Details of learning
- Stochastic Gradient Descent
  * Batch size 128
  * Momentum 0.9
  * Weight decay 0.0005

- Layer Initialization
  * Weight : zero-mean Gaussian distribution with standard deviation 0.01
  * Bias 
    * Constant 0 : conv 1, 3, fc 3
    * Constant 1 : conv 2, 4, 5 and fc 1, 2

- Learning Rate and Scheduler
  * Initial learning rate : 0.01
  * Divide by 10 when validation error rate stopped improving with current learning rate


## 6. Results
![image](https://user-images.githubusercontent.com/49643709/109617584-e947fb00-7b79-11eb-88e2-e1923a5fd437.png)
![image](https://user-images.githubusercontent.com/49643709/109617607-f06f0900-7b79-11eb-914a-18a3464b8faa.png)
- Models with asterisk * on ILSVRC-2012 : extra sixth convolutional layer over the last pooling layer
- 5 CNNs, 7 CNNs on ILSVRC-2012 : average the predictions


## 7. Discussion
- Deep convolutional neural network is capable of achieving record-breaking results on a highly challenging dataset using purely supervised learning
- Depth is important for achieving the results
